Replace circle CI with github actions

helenb opened the pull request:
This MR moves from circle ci over to github actions. I have leant heavily on the set-up for torchbox at https://github.com/torchbox/wagtail-torchbox/blob/main/.github/workflows/test.yml

Tasks the CI covers
- install npm dependencies
- install python and poetry
- lint front-end code
- compile front-end code
- lint back-end code
- test back-end code and create coverage report
- run isort, flake8 black and python tests.
- deployment to staging, dev, and prod

Note this MR also includes a couple of fixes to issues reported by flake8.

Further possible to dos (perhaps in separate merge requests):
- github pages for docs? (as per torchbox set up)
- add confirmation steps for deployment (requires using environments - see https://cloudlumberjack.com/posts/github-actions-approvals/)

helenb commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,68 @@
+
+    services:
+      postgres:
+        image: postgres:11.5
```
@kevinhowbrook this matches the version of postgres that was used in the circle ci tests, but not the version that I can see in heroku (I only have access to staging) or indeed in `docker-compose.yml` where it is different again.

kevinhowbrook replied:
yea it's a common (frustrating) thing that happens. I think for now match the docker compose file

kevinhowbrook commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,68 @@
+name: CI
+on: [push]
```
I think this might be better as:

```yml
on:
  pull_request:
  push:
    branches: [dev, staging, master]
```

The above means it will run on all PRs opened and pushes to the dev, staging and master

kevinhowbrook replied:
I think you could also delete the circleci folder too so it stop running as we will only use the github ci once done

helenb replied:
I was planning to delete the circleci folder once I'd finished referring to it - it's quite a useful reference for setting up the github actions for now.

helenb replied:
Should we also delete .gitlab-ci related files?

kevinhowbrook commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,68 @@
+      - name: black
+        run: black --check --diff ./rca fabfile.py --exclude migrations/
+      - name: collectstatic
+        run: python manage.py collectstatic --verbosity 0 --noinput --clear
```
Yea we do need collectstatic, looks like you got it in her though :)

helenb replied:
If I do manage to split them out would `black` and `flake8` count as `lint` (along with the npm tasks) or `test` in your view?

helenb replied:
I have now split them out - but it repeats the set up for python , poetry and npm across both which is maybe adding more time than is ideal.

kevinhowbrook commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,68 @@
+          poetry config virtualenvs.create false &&
+          poetry install
+      # - name: Flake8
+      #   run: flake8 ./rca fabfile.py
```
I think we need flake 8 to run.

As a sidenote, and something I think we should do in CI, you can run the entire pre-commit suite via:

```yml
jobs:
  precommit:
    name: Run precommit
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: 3.11

      - name: Set up Node.js
        uses: actions/setup-node@v4
        with:
          node-version: 20

      - run: npm ci --no-optional --no-audit --progress=false

      - name: Install pre-commit
        run: pip install pre-commit

      - name: Cache pre-commit dependencies
        uses: actions/cache@v3
        with:
          path: |
            ~/.cache/pre-commit
          key: pre-commit-${{ hashFiles('.pre-commit-config.yaml') }}

      - name: Run pre-commit
        run: pre-commit run --all-files
```

helenb replied:
Ah precommit is another area I am not too familiar with. I thought it was something that ran on your local machine, with certain tests that have to pass before you can commit code, rather than something that runs in CI? It's something I once tried years ago but haven't used since. Is it worth trying to set it up on my machine so I understand how it works?

kevinhowbrook replied:
It _is_ something that runs on your local machine yea, I'd highly recommend trying it out locally.
We are starting to use `pre-commit` in the CI too just because it ensures the following:
- All linting and formatting rules/versions are contained in pre-commit-config.yml
- It keeps the ci script a bit smaller when it comes to formatting and lining, we just just run the single pre-commit task.

It's not really a prerequestie of this PR but it's worth being aware of that as we are starting to bring it in more

helenb replied:
Are there some docs I can read somewhere on how to test it out?

kevinhowbrook reviewed (commented):
Shaping up well Helen, fab work. I've left a few comments around a few things but looks like you are sailing through it :)

kevinhowbrook commented:
Another thing to consider here is splitting out your actions, I think I mentioned that on the ticket but it might be worth doing before you look at deployments.

E.G, your `ci.yml` file could look like the following.
This would allow you to split the lining and testing out which is a bit more organised. It's not a standard thing, but sometimes nicer.

```yml
name: CI

on:
  pull_request:
  push:
    branches: [dev, staging, master]

jobs:
  lint:
    uses: ./.github/workflows/lint.yml

  test:
    uses: ./.github/workflows/test.yml
```

kevinhowbrook commented on .github/workflows/deploy-production.yaml:
```diff
@@ -0,0 +1,23 @@
+on:
+  push:
+    branches:
+      - main
```
I think that needs to be `master` as our branch still has that name

kevinhowbrook commented on .github/workflows/deploy-production.yaml:
```diff
@@ -0,0 +1,23 @@
+      - name: 'deploy master to heroku'
+        run: |
+          echo "Deploying to Heroku. To see progress, go to: https://dashboard.heroku.com/apps/$r/activity"
+          curl -sf -X POST -m 900 https://heroku-deploy.torchbox.com/$HEROKU_APP_NAME/$SHA?key=$DEPLOYMENT_KEY -d "Content-Length: 0"
```
Are you planning on adding the dev deployment as part of this too?

helenb replied:
Yep - I was thinking of having that in a different file. I'm not quite sure how best to test these deployment scripts.

helenb commented on .github/workflows/deploy-dev.yaml:
```diff
@@ -0,0 +1,19 @@
+      - uses: actions/checkout@v4
+      - name: 'deploy dev to heroku'
+        run: |
+          echo "Deploying to Heroku. To see progress, go to: https://dashboard.heroku.com/apps/$HEROKU_APP_NAME/activity"
```
when I tested this I couldn't actually see this output in the test shell

kevinhowbrook replied:
![image](https://github.com/torchbox/rca-wagtail-2019/assets/2128707/14f4619a-9c5d-4668-a944-0270ec441053)

helenb commented:
Hi @kevinhowbrook, I think this is nearly ready to merge now, and I won't have any more time for this for the considerable future, so I am hoping this is in a state now where you can take it back over and get it deployed.

There are two things I haven't done, which I don't think are essential:
- manual approval for deployments, which we need to use github environments for - I've added a link to a blog post about this in the MR description
- setting up the docs in github pages - I think this could be a fairly quick MR for someone else to do

Note that now I have removed the circle ci config, I am getting an error reported by circle ci on this branch - presumably removing the circle ci setup would be part of the deployment process for this branch?

I have added back in the flake8 test and fixed the errors it was reporting.

I have added in the coverage tests using the same bash script as circleci was using - it seems to run OK, but it includes a `$COVERAGE_TOKEN` which may need adding to the the repo secrets - I am not sure where this is coming from.
